### PR TITLE
Fixed #38, Check state root after processing block

### LINKED ISSUE #38: Check state root after processing block
Check the state root to make sure it's equal to the previous state root after processing the block.

### PR DESCRIPTION
**Pull request by wqking**

### CODE REVIEW

**Comment by meyer9** [comment]
Tests aren't passing. Can you make sure proposers are proposing the correct state root? I think it might need to be the previous state root actually.

**Comment by wqking** [comment]
I fixed the test, but seems there are conflict of how StateRoot is calculated.
In MineBlockWithSpecialsAndAttestations in util.go
```
	lastBlock, err := b.LastBlock()
	if err != nil {
		return nil, err
	}

	parentRoot, err := ssz.TreeHash(lastBlock)
	if err != nil {
		return nil, err
	}

	stateRoot, err := ssz.TreeHash(lastBlock)
```
State root is the hash of the block.

In GetStateRoot in beacon/rpc/rpc.go
```
func (s *server) GetStateRoot(ctx context.Context, in *empty.Empty) (*pb.GetStateRootResponse, error) {
	state := s.chain.GetState()

	root, err := ssz.TreeHash(state)
	if err != nil {
		return nil, err
	}

	return &pb.GetStateRootResponse{StateRoot: root[:]}, nil
}
```
State root is the hash of the state.

I think something wrong?

**Comment by meyer9** [comment]
Yeah, the state root calculation in util is wrong. It should get it from the blockchain, not the hash of the last block.

File: beacon/state.go [hunk 1]
```diff
@@ -69,6 +69,36 @@ func (b *Blockchain) ProcessBlock(block *primitives.Block, checkTime bool) error
...
+			if !found {
+				return errors.New("could not find state for block")
+			}
+
+			expectedStateRoot, err := ssz.TreeHash(expectedState)
+			if err != nil {
+				return err
+			}
+		*/
+		if block.BlockHeader.StateRoot != h {
```
Comment (meyer9) [inline beacon/state.go]: Don't compare hashes with == or !=. Use .IsEqual

**Comment by meyer9** [comment]
Move this to state.processBlock. There's a line starting with `TODO:`

**Comment by wqking** [comment]
How to access Blockchain instance in primitives.State.ProcessBlock? Do we need to pass Blockchain to the function?

**Comment by meyer9** [comment]
There should be a ChainView interface available for accessing the blockchain. Although, you might have to add a function to that interface for accessing state.

**Comment by meyer9** [comment]
type BlockView interface {

**Comment by wqking** [comment]
BlockView doesn't provide enough functions to verify the state root, such as `GetBlockByHash`, or access the state manager.
Indeed I wonder why to move the state root verification to primitives.State. primitives.State should not depend on the block chain, while the verification needs to be done by the blockchain.

**Comment by meyer9** [comment]
Hmm... I'd like to keep it all in the primitives package. You should add a function to the BlockView interface to access the state root for the previous slot.

**Comment by meyer9** [comment]
LGTM! I'll merge it once the tests pass.

**Pull request merged**